source/s3_client.c

```diff
@@ -487,26 +461,16 @@ static struct aws_s3_vip *s_s3_client_vip_new(
     return NULL;
 }
 
-/* Releases the memory for a vip structure. Assumes lock is held. */
+/* Releases the memory for a vip structure. */
 static void s_s3_client_vip_destroy(struct aws_s3_vip *vip) {
     AWS_PRECONDITION(vip);
 
+    aws_atomic_store_int(&vip->cleaning_up, 1);
+
     struct aws_s3_client *client = vip->owning_client;
     s_s3_client_lock_synced_data(client);
-
-    while (!aws_linked_list_empty(&vip->vip_connections)) {
```

Comment:
VIP's no longer hold a list of their VIP connections.  Instead, for clean up, the cleaning_up flag on the VIP will be retrieved in the work event loop, and they will be cleaned up there.